core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/dataset/DatasetResource.scala

```diff
@@ -1054,81 +1056,95 @@ class DatasetResource {
+            } finally {
+              zipOutputStream.closeEntry()
+            }
+          }
+        }.recover {
+          case e: IOException =>
+            throw new WebApplicationException("Error creating ZIP output stream", e)
+          case NonFatal(e) =>
+            throw new WebApplicationException("Unexpected error while creating ZIP", e)
+        }
       }
+    }
 
-      val targetDatasetPath = PathUtils.getDatasetPath(did)
-      val datasetVersion = getDatasetVersionByID(ctx, dvid)
-      val versionHash = datasetVersion.getVersionHash
-
-      val fileNodes = GitVersionControlLocalFileStorage.retrieveRootFileNodesOfVersion(
-        targetDatasetPath,
-        versionHash
+    Response
+      .ok(streamingOutput)
+      .header(
+        "Content-Disposition",
+        s"attachment; filename=${dataset.getName}-${version.getName}.zip"
       )
+      .`type`("application/zip")
+      .build()
+  }
 
-      val streamingOutput = new StreamingOutput() {
-        override def write(outputStream: OutputStream): Unit = {
-          val zipOutputStream = new ZipOutputStream(outputStream)
-
-          try {
-            fileNodes.foreach { fileNode =>
-              val zipEntryName = fileNode.getRelativePath.toString
-
-              try {
-                zipOutputStream.putNextEntry(new ZipEntry(zipEntryName))
-
-                val filePath = fileNode.getAbsolutePath
+  private def resolveAndValidatePath(
+      pathStr: String,
+      user: SessionUser
+  ): (Dataset, DatasetVersion) = {
+    val decodedPathStr = URLDecoder.decode(pathStr, StandardCharsets.UTF_8.name())
+    val (_, dataset, dsVersion, _) =
+      resolvePath(Paths.get(decodedPathStr), shouldContainFile = false)
 
-                Files.copy(filePath, zipOutputStream)
+    validateUserAccess(dataset.getDid, user.getUid)
+    (dataset, dsVersion)
+  }
 
-              } catch {
-                case e: IOException =>
-                  throw new WebApplicationException(s"Error processing file: $zipEntryName", e)
-              } finally {
-                zipOutputStream.closeEntry()
-              }
-            }
-          } catch {
-            case e: IOException =>
-              throw new WebApplicationException("Error creating ZIP output stream", e)
-          } finally {
-            zipOutputStream.close()
-          }
-        }
-      }
+  private def getLatestVersionInfo(did: UInteger, user: SessionUser): (Dataset, DatasetVersion) = {
+    validateUserAccess(did, user.getUid)
+    val dataset = getDatasetByID(context, did)
+    val latestVersion = getDatasetLatestVersion(context, did, user.getUid)
+    (dataset, latestVersion)
+  }
 
-      Response
-        .ok(streamingOutput)
-        .header(
-          "Content-Disposition",
-          s"attachment; filename=${dataset.getName}-${datasetVersion.getName}.zip"
-        )
-        .`type`("application/zip")
-        .build()
+  private def validateUserAccess(did: UInteger, uid: UInteger): Unit = {
```

Comment:
Remove this method.

Reply:
The method is needed as it is called twice and was in the original code.

Reply:
I see. Thanks.